chain/chain/src/store.rs

```diff
@@ -887,81 +887,72 @@ impl ChainStoreAccess for ChainStore {
         //         left working with a key that was used in the trie.
         //    2.2. Parse the trie key with a relevant KeyFor* implementation to ensure consistency
 
-        let data_key: Vec<u8> = match state_changes_request {
-            StateChangesRequest::AccountChanges { account_id } => {
-                utils::KeyForAccount::new(account_id).into()
-            }
-            StateChangesRequest::SingleAccessKeyChanges { account_id, access_key_pk } => {
-                utils::KeyForAccessKey::new(account_id, access_key_pk).into()
-            }
-            StateChangesRequest::AllAccessKeyChanges { account_id } => {
-                utils::KeyForAccessKey::get_prefix(account_id).into()
-            }
-            StateChangesRequest::CodeChanges { account_id } => {
-                utils::KeyForCode::new(account_id).into()
-            }
-            StateChangesRequest::DataChanges { account_id, key_prefix } => {
-                utils::KeyForData::new(account_id, key_prefix.as_ref()).into()
-            }
-        };
-        let storage_key = KeyForStateChanges::new(&block_hash, &data_key);
-
-        let mut changes_per_key_prefix = storage_key.find_iter(&self.store);
-
-        // It is a lifetime workaround. We cannot return `&mut changes_per_key_prefix.filter_map(...`
-        // as that is a temporary object created there with `filter_map`, and you cannot leak a
-        // reference to a temporary object.
-        let mut changes_per_exact_key;
-
-        let changes_per_key: &mut dyn Iterator<Item = _> = match state_changes_request {
-            // These request types are expected to match the key exactly:
-            StateChangesRequest::AccountChanges { .. }
-            | StateChangesRequest::SingleAccessKeyChanges { .. }
-            | StateChangesRequest::CodeChanges { .. } => {
-                changes_per_exact_key = changes_per_key_prefix.filter_map(|change| {
-                    let (key, state_changes) = match change {
-                        Ok(change) => change,
-                        error => {
-                            return Some(error);
-                        }
-                    };
-                    if key.len() != data_key.len() {
-                        None
-                    } else {
-                        debug_assert_eq!(key.as_ref(), data_key.as_ref() as &[u8]);
-                        Some(Ok((key, state_changes)))
-                    }
-                });
-                &mut changes_per_exact_key
-            }
-
-            StateChangesRequest::AllAccessKeyChanges { .. }
-            | StateChangesRequest::DataChanges { .. } => &mut changes_per_key_prefix,
-        };
-
         Ok(match state_changes_request {
-            StateChangesRequest::AccountChanges { account_id, .. } => {
-                StateChanges::from_account_changes(changes_per_key, account_id)?
+            StateChangesRequest::AccountChanges { account_ids } => {
```

Comment:
FYI, I finally arrived at the state when the code for reading the state changes from the storage and converting those raw values into proper responses is done outside of this function in the right places, so ultimately, it is easier to just "monomorphize" the implementations for each individual type of request (notice how the diff size is +95 -90 meaning that implementing a new feature only required a few extra lines of code).

Reply:
Well, the PR progressed to +727 -520 :rofl: Well, those are mostly tests, which now cover not only happy paths but also corner cases.